File: analyzer/fen_parser.py
def fen_to_vector(fen):
    """Convert FEN to 769-dimensional input vector:
    - 64 squares × 12 piece types (one-hot) = 768
    - 1 for turn (white=1, black=0)
    """
    parts = fen.strip().split()
    if len(parts) < 2:
        raise ValueError(f"Invalid FEN: {fen}")
    
    board_part = parts[0]
    turn = parts[1]
    
    piece_map = {
        'P': 0, 'N': 1, 'B': 2, 'R': 3, 'Q': 4, 'K': 5,
        'p': 6, 'n': 7, 'b': 8, 'r': 9, 'q': 10, 'k': 11
    }
    
    vector = [0.0] * 769
    
    square = 0
    for char in board_part:
        if char == '/':
            continue
        elif char.isdigit():
            square += int(char)
        else:
            if char in piece_map:
                piece_idx = piece_map[char]
                vector[square * 12 + piece_idx] = 1.0
            square += 1
    
    vector[768] = 1.0 if turn == 'w' else 0.0
    
    return vector

def label_to_vector(label):
    """Convert label to one-hot vector: Nothing, Check, Checkmate, Stalemate
    Handles labels with optional color suffix (e.g., 'Checkmate White')"""
    label = label.strip().lower()
    
    # Extract base label (ignore color suffix)
    base_label = label.split()[0] if ' ' in label else label
    
    labels = {"nothing": 0, "check": 1, "checkmate": 2, "stalemate": 3}
    
    if base_label not in labels:
        raise ValueError(f"Invalid label: {label}")
    
    vector = [0.0] * 4
    vector[labels[base_label]] = 1.0
    return vector

def vector_to_label(vector):
    """Convert output vector to label"""
    labels = ["Nothing", "Check", "Checkmate", "Stalemate"]
    return labels[vector.index(max(vector))]

File: analyzer_cpp/fen_parser.cpp
#include "fen_parser.hpp"
#include <sstream>
#include <stdexcept>
#include <map>
#include <algorithm>
#include <cctype>

static std::string to_lower(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(), ::tolower);
    return s;
}

std::vector<double> fen_to_vector(const std::string& fen) {
    std::istringstream iss(fen);
    std::string board_part, turn;
    iss >> board_part >> turn;
    
    if (board_part.empty() || turn.empty()) {
        throw std::runtime_error("Invalid FEN: " + fen);
    }
    
    std::map<char, int> piece_map = {
        {'P', 0}, {'N', 1}, {'B', 2}, {'R', 3}, {'Q', 4}, {'K', 5},
        {'p', 6}, {'n', 7}, {'b', 8}, {'r', 9}, {'q', 10}, {'k', 11}
    };
    
    std::vector<double> vec(769, 0.0);
    
    int square = 0;
    for (char c : board_part) {
        if (c == '/') {
            continue;
        } else if (std::isdigit(c)) {
            square += (c - '0');
        } else {
            auto it = piece_map.find(c);
            if (it != piece_map.end()) {
                vec[square * 12 + it->second] = 1.0;
            }
            square++;
        }
    }
    
    vec[768] = (turn == "w" || turn == "W") ? 1.0 : 0.0;
    
    return vec;
}

std::vector<double> label_to_vector(const std::string& label) {
    std::string lower_label = to_lower(label);
    
    std::map<std::string, int> labels = {
        {"nothing", 0},
        {"check white", 1},
        {"check black", 2},
        {"checkmate white", 3},
        {"checkmate black", 4},
        {"stalemate", 5}
    };
    
    auto it = labels.find(lower_label);
    if (it == labels.end()) {
        throw std::runtime_error("Invalid label: " + label);
    }
    
    std::vector<double> vec(6, 0.0);
    vec[it->second] = 1.0;
    return vec;
}

std::string vector_to_label(const std::vector<double>& vec) {
    std::vector<std::string> labels = {"Nothing", "Check White", "Check Black", "Checkmate White", "Checkmate Black", "Stalemate"};
    int max_idx = 0;
    for (size_t i = 1; i < vec.size(); i++) {
        if (vec[i] > vec[max_idx]) {
            max_idx = i;
        }
    }
    return labels[max_idx];
}
